cli/types/index.d.ts

```diff
@@ -811,14 +811,10 @@ declare namespace Cypress {
 
     /**
      * Invoke a function on the previously yielded subject.
-     * This isn't possible to strongly type without generic override yet.
-     * If called on an object you can do this instead: `.then(s => s.show())`.
-     * If called on an array you can do this instead: `.each(s => s.show())`.
-     * From there the subject will be properly typed.
      *
      * @see https://on.cypress.io/invoke
      */
-    invoke(functionName: keyof Subject, ...args: any[]): Chainable<Subject> // don't have a way to express return types yet
+    invoke<K extends { [K in keyof Subject]: Subject[K] extends (...args: any[]) => any ? K : never }[keyof Subject], R extends Subject[K] extends (...args: any[]) => infer R ? R : never>(functionName: K, ...args: any[]): Chainable<R>

```

Comment:
Thus also argument types can be inferred like so
```ts
    invoke<
     K extends { [K in keyof Subject]: Subject[K] extends (...args: any) => any ? K : never }[keyof Subject],
     R extends Subject[K] extends (...args: any[]) => infer R ? R : never,
     A extends Subject[K] extends (...args: infer A) => any ? A : never,
    >(functionName: K, ...args: A): Chainable<R>
```

Then it will fail if 
```ts
cy.wrap({ foo: () => 1, bar: 2 }).invoke('foo', '15').should('equal', 1)
```

And also I don't know about typescript definitions practices here, but for such complex generics it is always good practice to name types like `TFunctionName` `TReturnType`. Because right now it is really hard to understand what each parameter mean

Reply:
Thanks for the feedback; I've updated accordingly.
In order to infer the args correctly, we'll have to use TypeScript 3.0+, which is reflected in the latest commit